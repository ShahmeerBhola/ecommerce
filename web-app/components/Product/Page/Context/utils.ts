import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
