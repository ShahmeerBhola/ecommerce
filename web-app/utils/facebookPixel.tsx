import React from 'react';
import config from '../config';

const { facebookPixelID } = config;

// https://www.facebook.com/business/help/402791146561655?id=1205376682832142
const event = <P,>(eventName: string, params?: P): void => {
  if (facebookPixelID !== null) {
    if (params !== undefined) {
      window.fbq('track', eventName, params);
    } else {
      window.fbq('track', eventName);
    }
  }
};

export const addLineItemToCart = ({ sku, name, price }: WordPress.ProductBase, type = 'line_item'): void =>
  event<facebook.Pixel.AddToCartParameters>('AddToCart', {
    content_name: name,
    content_type: type,
    content_ids: [sku],
    value: price,
    currency: config.defaultCurrencyCode,
  });

export const addWheelTirePackageToCart = ({ wheel, tire, addOns }: Cart.WheelTirePackage): void => {
  const addWheelTirePackageItemToCart = (product: WordPress.ProductBase): void =>
    addLineItemToCart(product, 'wheel_tire_package');

  addWheelTirePackageItemToCart(wheel);
  addWheelTirePackageItemToCart(tire);
  addOns.map((addOn) => addWheelTirePackageItemToCart(addOn));
};

export const initiateCheckout = (): void => event('InitiateCheckout');
export const completePurchase = (params: facebook.Pixel.PurchaseParameters): void =>
  event<facebook.Pixel.PurchaseParameters>('Purchase', params);

export const FacebookPixelScripts: React.FunctionComponent = () => {
  {
    if (facebookPixelID !== null) {
      return (
        <>
          <script
            dangerouslySetInnerHTML={{
              __html: `
            !function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod ?
              n.callMethod.apply(n, arguments) : n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
 fbq('init', '${facebookPixelID}');
fbq('track', 'PageView');
          `,
            }}
          />
          <noscript>
            <img
              height="1"
              width="1"
              src={`https://www.facebook.com/tr?id=${facebookPixelID}&ev=PageView
&noscript=1`}
            />
          </noscript>
        </>
      );
    }
  }
  return null;
};
