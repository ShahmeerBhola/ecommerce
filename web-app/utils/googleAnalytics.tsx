import React from 'react';
import config from '../config';
import { isBrowserSide } from './';

const { googleAnalyticsTrackingID, googleAdsConversionID } = config;

// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
// https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics
export const pageview = (url: string): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('config', googleAnalyticsTrackingID, {
      page_path: url,
    });
  }
};

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
const event = (action: string, eventParams: any): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('event', action, eventParams);
  }
};

const conversionEvent = (sendTo: string, eventParams?: StringMap): void =>
  event('conversion', { send_to: sendTo, ...eventParams });

export const sendOrderConfirmationConversionEvent = (orderId: string): void =>
  conversionEvent('AW-479563831/zZZcCPqr3-gBELeg1uQB', { transaction_id: orderId });

export const sendContactFormConversionEvent = (): void => conversionEvent('AW-479563831/6i-DCKyL6-gBELeg1uQB');

export const sendRequestForQuoteEvent = (): void => conversionEvent('AW-479563831/C9PsCJv7_-gBELeg1uQB');

// https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce
export const sendEnhancedEcommerceEvent = (
  eee: GoogleAnalytics.EnhancedEcommerceEvent,
  items: GoogleAnalytics.Product[],
): void => event(eee, { items });

export const sendEnhancedEcommercePurchaseEvent = (data: GoogleAnalytics.PurchaseEvent): void =>
  event('purchase', data);

export const sendEnhancedEcommerceCheckoutOptionEvent = (data: GoogleAnalytics.CheckoutProgressEvent): void =>
  event('set_checkout_option', data);

export const enhancedEcommerceLineItemListName = 'Line Items';
export const enhancedEcommerceWheelTirePackageListName = 'Wheel Tire Package';
export const enhancedEcommerceWheelTirePackageAddOnsListName = 'Wheel Tire Package Add Ons';

export const GoogleAnalyticsAndAdsScripts: React.FunctionComponent = () => {
  {
    if (googleAnalyticsTrackingID !== null && googleAdsConversionID !== null) {
      return (
        <>
          <script async src={`https://www.googletagmanager.com/gtag/js?id=${googleAnalyticsTrackingID}`} />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${googleAnalyticsTrackingID}', {
              page_path: window.location.pathname,
            });
            gtag('config', '${googleAdsConversionID}')
          `,
            }}
          />
        </>
      );
    }
  }
  return null;
};
