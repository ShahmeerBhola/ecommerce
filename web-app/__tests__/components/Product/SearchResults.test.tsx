import React from 'react';
import renderer from 'react-test-renderer';
import { generateMockProduct, generateMockPagination } from '../../utils';
import SearchResults from '../../../components/Product/SearchResults';

describe('SearchResults component', () => {
  const products = [generateMockProduct(1), generateMockProduct(2), generateMockProduct(3)];
  const pagination = generateMockPagination(3, 1, 1, 25);

  test('it renders properly', () => {
    const tree = renderer.create(<SearchResults products={products} pagination={pagination} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
