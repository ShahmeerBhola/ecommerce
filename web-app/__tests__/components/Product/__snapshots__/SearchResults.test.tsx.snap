// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults component it renders properly 1`] = `
Array [
  <hr
    className="my-10 w-full mx-auto border-1 border-solid border-gray-100"
  />,
  <div
    className="flex flex-wrap"
  >
    <div
      className="w-full sm:w-1/2 lg:w-1/4 p-5"
    >
      <div
        className="bg-white p-5"
      >
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-1"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=""
          >
            <img
              alt=""
              className="w-full h-full"
              src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
            />
          </div>
        </a>
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-1"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p
            className="font-secondary font-normal text-black text-base my-5 uppercase font-bold"
          >
            American Force Battery CC
            ...
          </p>
        </a>
        <div
          className="flex items-center"
        >
          <div
            className="w-1/2"
          >
            <p
              className="font-secondary font-medium text-red-100 text-base"
            >
              $11,536.00
            </p>
          </div>
          <div
            className="w-1/2 text-right"
          >
            <a
              className="inline-block mr-3"
              href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-1"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="48"
                  stroke="#F23939"
                  strokeWidth="2"
                  width="48"
                  x="1"
                  y="1"
                />
                <path
                  d="M30.0168 31L28.2218 29.2L31.201 26.2125L14 26.2125L14 23.6669L31.0807 23.6669L28.2218 20.8L30.0168 19L36 25L30.0168 31Z"
                  fill="#F23939"
                />
              </svg>
            </a>
            <div
              className="cursor-pointer inline-block"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#F23939"
                  height="50"
                  width="50"
                />
                <path
                  d="M18.325 32.5875H30.3661L31.1103 29.5588H15.3743L12 15.8266H34.4845L35.6705 11H42V12.7979H37.0503L32.1699 32.6593C33.5906 32.9697 34.6591 34.2569 34.6591 35.7938C34.6591 37.5617 33.2454 39 31.5077 39C29.77 39 28.3563 37.5616 28.3563 35.7938C28.3563 35.2887 28.472 34.8108 28.6774 34.3854H21.1553C21.3607 34.8108 21.4764 35.2887 21.4764 35.7938C21.4764 37.5617 20.0627 39 18.325 39C16.5874 39 15.1736 37.5616 15.1736 35.7938C15.1737 34.0259 16.5874 32.5875 18.325 32.5875Z"
                  fill="white"
                />
                <path
                  d="M23.7656 19H26.2656V27H23.7656V19Z"
                  fill="#F23939"
                />
                <path
                  d="M29 21.7383L29 24.2383L21 24.2383L21 21.7383L29 21.7383Z"
                  fill="#F23939"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="w-full sm:w-1/2 lg:w-1/4 p-5"
    >
      <div
        className="bg-white p-5"
      >
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=""
          >
            <img
              alt=""
              className="w-full h-full"
              src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
            />
          </div>
        </a>
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p
            className="font-secondary font-normal text-black text-base my-5 uppercase font-bold"
          >
            American Force Battery CC
            ...
          </p>
        </a>
        <div
          className="flex items-center"
        >
          <div
            className="w-1/2"
          >
            <p
              className="font-secondary font-medium text-red-100 text-base"
            >
              $11,536.00
            </p>
          </div>
          <div
            className="w-1/2 text-right"
          >
            <a
              className="inline-block mr-3"
              href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="48"
                  stroke="#F23939"
                  strokeWidth="2"
                  width="48"
                  x="1"
                  y="1"
                />
                <path
                  d="M30.0168 31L28.2218 29.2L31.201 26.2125L14 26.2125L14 23.6669L31.0807 23.6669L28.2218 20.8L30.0168 19L36 25L30.0168 31Z"
                  fill="#F23939"
                />
              </svg>
            </a>
            <div
              className="cursor-pointer inline-block"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#F23939"
                  height="50"
                  width="50"
                />
                <path
                  d="M18.325 32.5875H30.3661L31.1103 29.5588H15.3743L12 15.8266H34.4845L35.6705 11H42V12.7979H37.0503L32.1699 32.6593C33.5906 32.9697 34.6591 34.2569 34.6591 35.7938C34.6591 37.5617 33.2454 39 31.5077 39C29.77 39 28.3563 37.5616 28.3563 35.7938C28.3563 35.2887 28.472 34.8108 28.6774 34.3854H21.1553C21.3607 34.8108 21.4764 35.2887 21.4764 35.7938C21.4764 37.5617 20.0627 39 18.325 39C16.5874 39 15.1736 37.5616 15.1736 35.7938C15.1737 34.0259 16.5874 32.5875 18.325 32.5875Z"
                  fill="white"
                />
                <path
                  d="M23.7656 19H26.2656V27H23.7656V19Z"
                  fill="#F23939"
                />
                <path
                  d="M29 21.7383L29 24.2383L21 24.2383L21 21.7383L29 21.7383Z"
                  fill="#F23939"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="w-full sm:w-1/2 lg:w-1/4 p-5"
    >
      <div
        className="bg-white p-5"
      >
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-3"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=""
          >
            <img
              alt=""
              className="w-full h-full"
              src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
            />
          </div>
        </a>
        <a
          href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-3"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <p
            className="font-secondary font-normal text-black text-base my-5 uppercase font-bold"
          >
            American Force Battery CC
            ...
          </p>
        </a>
        <div
          className="flex items-center"
        >
          <div
            className="w-1/2"
          >
            <p
              className="font-secondary font-medium text-red-100 text-base"
            >
              $11,536.00
            </p>
          </div>
          <div
            className="w-1/2 text-right"
          >
            <a
              className="inline-block mr-3"
              href="/product/ck17-3016-5x500-sf-american-force-battery-cc-ck17-30x16-5x5-3"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="48"
                  stroke="#F23939"
                  strokeWidth="2"
                  width="48"
                  x="1"
                  y="1"
                />
                <path
                  d="M30.0168 31L28.2218 29.2L31.201 26.2125L14 26.2125L14 23.6669L31.0807 23.6669L28.2218 20.8L30.0168 19L36 25L30.0168 31Z"
                  fill="#F23939"
                />
              </svg>
            </a>
            <div
              className="cursor-pointer inline-block"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="50"
                viewBox="0 0 50 50"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#F23939"
                  height="50"
                  width="50"
                />
                <path
                  d="M18.325 32.5875H30.3661L31.1103 29.5588H15.3743L12 15.8266H34.4845L35.6705 11H42V12.7979H37.0503L32.1699 32.6593C33.5906 32.9697 34.6591 34.2569 34.6591 35.7938C34.6591 37.5617 33.2454 39 31.5077 39C29.77 39 28.3563 37.5616 28.3563 35.7938C28.3563 35.2887 28.472 34.8108 28.6774 34.3854H21.1553C21.3607 34.8108 21.4764 35.2887 21.4764 35.7938C21.4764 37.5617 20.0627 39 18.325 39C16.5874 39 15.1736 37.5616 15.1736 35.7938C15.1737 34.0259 16.5874 32.5875 18.325 32.5875Z"
                  fill="white"
                />
                <path
                  d="M23.7656 19H26.2656V27H23.7656V19Z"
                  fill="#F23939"
                />
                <path
                  d="M29 21.7383L29 24.2383L21 24.2383L21 21.7383L29 21.7383Z"
                  fill="#F23939"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <hr
    className="my-10 w-full mx-auto border-1 border-solid border-gray-100"
  />,
  <div
    className="inline-flex items-center float-right"
  >
    <p
      className="font-secondary font-normal text-gray-100 text-base mr-5"
    >
      Page
    </p>
    <div
      className="mx-2 py-2 px-3 cursor-pointer bg-white border border-gray-80 border-solid"
      onClick={[Function]}
    >
      <p
        className="font-secondary font-normal text-red-100 text-base"
      >
        1
      </p>
    </div>
  </div>,
]
`;
